src/qibo/transpiler/router.py

```diff
@@ -664,13 +667,22 @@ def __call__(self, circuit: Circuit, initial_layout: dict):
             (:class:`qibo.models.circuit.Circuit`, dict): routed circuit and final layout.
         """
         self._preprocessing(circuit=circuit, initial_layout=initial_layout)
+        self._saved_circuit = deepcopy(self.circuit)
+        longest_path = np.max(self._dist_matrix)
+
         while self._dag.number_of_nodes() != 0:
             execute_block_list = self._check_execution()
             if execute_block_list is not None:
                 self._execute_blocks(execute_block_list)
             else:
                 self._find_new_mapping()
 
+            # If the number of added swaps is too high, the algorithm is stuck.
+            # Reset the circuit to the last saved state and make the nearest gate executable by manually adding SWAPs.
+            if self._added_swaps > 1.5 * longest_path:  # threshold is arbitrary
```

Comment:
We can leave this treshold as a parameter (in the __init__) that can be defined by the user with 1.5 as default value.